lib/widgets/message_list.dart

```diff
@@ -530,6 +534,68 @@ class MarkAsReadWidget extends StatelessWidget {
   }
 }
 
+class _TypingStatusState extends State<TypingStatusWidget> with PerAccountStoreAwareStateMixin<TypingStatusWidget> {
+  TypingStatus? model;
+
+  @override
+  void onNewStore() {
+    model?.removeListener(_modelChanged);
+    model = PerAccountStoreWidget.of(context).typingStatus
+      ..addListener(_modelChanged);
+  }
+
+  @override
+  void dispose() {
+    model?.removeListener(_modelChanged);
+    super.dispose();
+  }
+
+  void _modelChanged() {
+    setState(() {
+      // The actual state lives in [model].
+      // This method was called because that just changed.
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+    final narrow = widget.narrow;
+    const placeholder = SizedBox(height: 8);
+    if (narrow is! SendableNarrow) return placeholder;
+
+    final localization = ZulipLocalizations.of(context);
+    final typistNames = model!.typistIdsInNarrow(narrow)
+      .where((id) => id != store.selfUserId)
+      .map((id) => store.users[id]?.fullName ?? localization.unknownUserName)
+      .toList();
+    if (typistNames.isEmpty) return placeholder;
+
+    final String text = switch (typistNames.length) {
```

Comment:
Specifically, once there's no `where` here, these two length checks can happily use the original Iterable instead of the result of the `map`.

Then we turn out not to need the List at all; and if there are 3+ typists in the list, we don't need to go and look up each of their names, which we're not going to use.

Reply:
I'm not sure if I understand "these two length checks can happily use the original Iterable instead of the result of the map." here. Do you mean something like this?

```dart
    final String text = switch (typistIds) {
      [final firstTypist] =>
        localizations.onePersonTyping(store.users[firstTypist]?.fullName ?? localizations.unknownUserName),
      [final firstTypist, final secondTypist] =>
        localizations.twoPeopleTyping(
          store.users[firstTypist]?.fullName ?? localizations.unknownUserName,
          store.users[secondTypist]?.fullName ?? localizations.unknownUserName,
        ),
      _ => localizations.manyPeopleTyping,
    };
```

Reply:
Like that except with no `toList`.

Which I think means the `switch` will be on `typistIds.length`, rather than `typistIds` itself.